src/message/dkim.rs

```diff
@@ -0,0 +1,316 @@
+use crate::message::{header::HeaderName, Headers, Message};
+use base64::encode;
+use sha2::{Sha256, Digest};
+use regex::Regex;
+use rsa::{pkcs1::FromRsaPrivateKey, Hash, PaddingScheme, RsaPrivateKey};
+use std::fmt::Display;
+use std::time::SystemTime;
+
+/// Describe Dkim Canonicalization to apply to either body or headers
+#[derive(Copy, Clone)]
+pub enum DkimCanonicalizationType {
+    Simple,
+    Relaxed,
+}
+
+impl Display for DkimCanonicalizationType {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
+        match self {
+            DkimCanonicalizationType::Simple => write!(fmt, "simple"),
+            DkimCanonicalizationType::Relaxed => write!(fmt, "relaxed"),
+        }
+    }
+}
+
+/// Describe Canonicalization to be applied before signing
+#[derive(Copy, Clone)]
+pub struct DkimCanonicalization {
+    header: DkimCanonicalizationType,
+    body: DkimCanonicalizationType,
+}
+
+/// Format canonicalization to be shown in Dkim header
+impl Display for DkimCanonicalization {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
+        write!(fmt, "{}/{}", self.header, self.body)
+    }
+}
+
+/// A struct to describe Dkim configuration applied when signing a message
+/// selector: the name of the key publied in DNS
+/// domain: the domain for which we sign the message
+/// private_key: private key in PKCS1 string format
+/// headers: a list of headers name to be included in the signature. Signing of more than one
+/// header with same name is not supported
+/// canonicalization: the canonicalization to be applied on the message
+#[derive(Clone)]
+pub struct DkimConfig {
```

Comment:
This seems to currently be an `RSA`-only DKIM configuration and implementation. Even is we don't have an alternative for now in lettre, we may want to support [Ed25519](https://datatracker.ietf.org/doc/html/rfc8463) too later.

Reply:
Indeed, I didn't notice the update in the RFC for signing algorithms. It shouldn't be long to implement with an Enum in the config for the key type

Reply:
For the hash part, sha256 is currently the only accepted algorithm in the RFCs so it's probably fine to hardcode it.